File: apps/mobile/src/features/foods/utils/foodDuplicateName.ts
const GENERATED_COPY_SUFFIX = /^(.+) Copy(?: ([1-9][0-9]*))?$/u;
const MAX_GENERATED_ORDINAL = Number.MAX_SAFE_INTEGER - 1;

export function allocateDuplicateFoodName(
  sourceName: string,
  activeNames: readonly string[],
  sourceIsDuplicate: boolean,
): string {
  let baseName = sourceName;
  let startingOrdinal = 1;

  if (sourceIsDuplicate) {
    const match = GENERATED_COPY_SUFFIX.exec(sourceName);
    if (match) {
      const parsed = Number.parseInt(match[2] ?? "1", 10);
      if (Number.isSafeInteger(parsed) && parsed >= 1 && parsed <= MAX_GENERATED_ORDINAL) {
        baseName = match[1]!;
        startingOrdinal = parsed + 1;
      }
    }
  }

  const usedNames = new Set(activeNames);
  for (let ordinal = startingOrdinal; ordinal <= Number.MAX_SAFE_INTEGER; ordinal += 1) {
    const candidate = ordinal === 1 ? `${baseName} Copy` : `${baseName} Copy ${ordinal}`;
    if (!usedNames.has(candidate)) return candidate;
  }

  throw new Error("Duplicate Food name suffix exhausted the supported range.");
}

File: apps/backend/app/domain/food_duplicate_name.py
from __future__ import annotations

import re
from collections.abc import Iterable

_GENERATED_COPY_SUFFIX = re.compile(r"^(?P<base>.+) Copy(?: (?P<ordinal>[1-9][0-9]*))?$")
_MAX_GENERATED_ORDINAL = 9_007_199_254_740_990


def allocate_duplicate_food_name(
    source_name: str,
    active_names: Iterable[str],
    *,
    source_is_duplicate: bool,
) -> str:
    base_name = source_name
    starting_ordinal = 1

    if source_is_duplicate:
        match = _GENERATED_COPY_SUFFIX.fullmatch(source_name)
        if match is not None:
            parsed_ordinal = int(match.group("ordinal") or "1")
            if parsed_ordinal <= _MAX_GENERATED_ORDINAL:
                base_name = match.group("base")
                starting_ordinal = parsed_ordinal + 1

    used_names = set(active_names)
    ordinal = starting_ordinal
    while ordinal <= _MAX_GENERATED_ORDINAL + 1:
        candidate = f"{base_name} Copy" if ordinal == 1 else f"{base_name} Copy {ordinal}"
        if candidate not in used_names:
            return candidate
        ordinal += 1
    raise ValueError("Duplicate Food name suffix exhausted the supported range.")
